### PR TITLE
out_azure: support using timestamp as timegenerated

### LINKED ISSUE #1423: Support time-generated-field for Azure
Allow to specify a field that will be used as `TimeGenerated` timestamp.
To do that a new HTTP header [`time-generated-field`](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#request-headers) must be added.

### PR DESCRIPTION
**Pull request by tobiaskohlbau**
When supplied the timegenerated field can override the automatic generated
ingestion timestamp from azure. In order to support this usecase a new option
time_generated is introduced. If this setting gets enabled the already
present timestamp value will be formatted in accordance with the azure api
specification and during the api call the setting to override the
timegenerated field gets set.

The documentation is available at:
https://docs.microsoft.com/en-us/azure/azure-monitor/logs/data-collector-api

Fixes #1423

----
**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [N/A] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

### Example
```
[INPUT]
    Name   dummy
    Tag    source.docker.test
    Rate 1000

[OUTPUT]
    Name        azure
    Match       source.docker.*
    Customer_ID ID_GOES_HERE
    Shared_Key  SHARED_KEY_HERE
    Log_Type test
    Time_Generated On
```

### Valgrind
```
==22492==
==22492== HEAP SUMMARY:
==22492==     in use at exit: 108,278 bytes in 3,682 blocks
==22492==   total heap usage: 247,268 allocs, 243,586 frees, 1,107,529,898 bytes allocated
==22492==
==22492== LEAK SUMMARY:
==22492==    definitely lost: 0 bytes in 0 blocks
==22492==    indirectly lost: 0 bytes in 0 blocks
==22492==      possibly lost: 0 bytes in 0 blocks
==22492==    still reachable: 108,278 bytes in 3,682 blocks
==22492==         suppressed: 0 bytes in 0 blocks
==22492== Reachable blocks (those to which a pointer was found) are not shown.
==22492== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==22492==
==22492== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Documentation**

- [x] Documentation required for this feature

I will update the docs if this PR gets accepted in it's current state.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by tobiaskohlbau** [comment]
By the looks of it the commit checker is wrong. The regex should allow `_` as described as valid prefix within https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#commit-changes.

https://github.com/fluent/fluent-bit/blob/0b84150719a0b563f7f19f9d644a40af87a97bc1/.github/workflows/check-commit.yaml#L17

Should be:
```
pattern: '^[a-z\-_]+\:[ ]{0,1}[a-z]+[a-zA-Z0-9 \-\.\:]+$'
```

**Comment by tobiaskohlbau** [comment]
I've updated the PR to the latest _master_ branch. Is there a todo for me at this point?

**Comment by tobiaskohlbau** [comment]
I kindly ask for some information what is holding this PR from getting a review or a comment that this is not intended for the upstream repo.

**Comment by tobiaskohlbau** [comment]
I'm still waiting for one of the maintainers to let me know the further process to get this upstreamed.

**Comment by seanscoggins** [comment]
Would really like to see this merged - what is needed and how can I help?

**Comment by tobiaskohlbau** [comment]
I know maintainers are sometimes busy and I'm totally fine with that. Is it at least possible for anyone with the correct rights to disable the stale bot? I would like to avoid spamming this PR every 45 days in order to keep it from being autoclosed.

**Comment by agup006** [comment]
@nokute78 maybe you could take a look?

File: plugins/out_azure/azure.c
Comment (nokute78) [inline plugins/out_azure/azure.c:108]: https://docs.microsoft.com/en-us/azure/azure-monitor/logs/data-collector-api#request-headers
```
" The contents of the message field should follow the ISO 8601 format YYYY-MM-DDThh:mm:ssZ."
```

Do we need last character 'Z' ?
`FLB_PACK_JSON_DATE_ISO8601` is `"%Y-%m-%dT%H:%M:%S"`
The format string doesn't contain last 'Z'.

File: plugins/out_azure/azure.c
Comment (nokute78) [inline plugins/out_azure/azure.c:113]: Does the azure API support microsecond ?

https://docs.microsoft.com/en-us/azure/azure-monitor/logs/data-collector-api#request-headers
```
" The contents of the message field should follow the ISO 8601 format YYYY-MM-DDThh:mm:ssZ."
```

File: plugins/out_azure/azure.c
Comment (nokute78) [inline plugins/out_azure/azure.c:106]: Is `gmtime_r` supported on Windows system ?  Since we have to support to build fluent-bit on Windows.

https://linux.die.net/man/3/gmtime_r
```
Thread-safe versions asctime_r(), ctime_r(), gmtime_r() and localtime_r() are specified by SUSv2, and available since libc 5.2.5.
```
I think it is an unix api.

Hmm, Appvayor is succeeded, so Windows supports it?

File: plugins/out_azure/azure_conf.c
Comment (nokute78) [inline plugins/out_azure/azure_conf.c:121]: We support `flb_utils_bool` function.
Could you use the API ?

File: plugins/out_azure/azure_conf.c
Comment (nokute78) [inline plugins/out_azure/azure_conf.c:253]: Could you remove this function ?
We already support `flb_utils_bool` function.

**Review by nokute78 — CHANGES_REQUESTED** [review]
I added a general request/question comments since I'm not familiar with azure...
So, could you attach a picture or log or something to confirm azure approving the request ?

File: plugins/out_azure/azure.c
Comment (tobiaskohlbau) [reply plugins/out_azure/azure.c:108]: Azure is ingesting this format fine at the moment. I'm running the current Patch for a few months now. But I think the best road to go down with is do comply with the documentation. I will update the Patch.

File: plugins/out_azure/azure.c
Comment (tobiaskohlbau) [reply plugins/out_azure/azure.c:113]: Looks like they do. I've checked the timestamp from the log and the timestamp reported in Azure.

Timestamp in for e.g. docker: 2021-08-05T08:06:48.277911316Z
Timegenerated in Azure: 2021-08-05T08:06:48.277Z

I can't remember where I found the information but during implementation I've gathered that they support milliseconds.

File: plugins/out_azure/azure_conf.c
Comment (tobiaskohlbau) [reply plugins/out_azure/azure_conf.c:121]: Ofcourse! I wasn't aware of that API. I've digged through the _influxdb.c_ sourcecode and found it there. Looks like it already got fixed in influx as well. https://github.com/fluent/fluent-bit/commit/11323419cc060fbdf4546903fc52289a1343a58f#diff-3b3a56855868026d54acd7df3827545b9e3ecc04c4cc8b44239289e4c99d550d

File: plugins/out_azure/azure.c
Comment (tobiaskohlbau) [reply plugins/out_azure/azure.c:106]: fluent-bit has a compatibility layer: https://github.com/fluent/fluent-bit/blob/33f80ddc85443316fabd9877f4c0024956fe1e36/include/fluent-bit/flb_compat.h#L73-L79

File: plugins/out_azure/azure.c
Comment (tobiaskohlbau) [reply plugins/out_azure/azure.c:108]: Sorry for the confusion the Z is already appended later after the milliseconds. See https://github.com/tobiaskohlbau/fluent-bit/blob/791f388274e0075c6bb487f4240a2ed343f8948b/plugins/out_azure/azure.c#L112

**Comment by tobiaskohlbau** [comment]
> So, could you attach a picture or log or something to confirm azure approving the request ?

Do you mean approving this PR or ingesting the data? This PR is running in my environment for roughly 6 months now without an issue.

File: plugins/out_azure/azure.c
Comment (nokute78) [reply plugins/out_azure/azure.c:106]: Thank you for pointing it.

File: plugins/out_azure/azure.c
Comment (nokute78) [reply plugins/out_azure/azure.c:108]: Thank you.

File: plugins/out_azure/azure_conf.c
Comment (nokute78) [reply plugins/out_azure/azure_conf.c]: Thank you.

**Comment by nokute78** [comment]
> Do you mean approving this PR or ingesting the data? This PR is running in my environment for roughly 6 months now without an issue.

I want to merge the PR.
However I don't have test environment so I need something as an evidence to merge.

Of course we don't need confidential information.
Please remove/hide it :)

File: plugins/out_azure/azure_conf.c
Comment (nokute78) [reply plugins/out_azure/azure_conf.c:248]: Please remove this function to reduce maintenance cost.

Comment (tobiaskohlbau) [reply plugins/out_azure/azure_conf.c:248]: I'm sorry I thought it was deleted. Must be in staging phase 🙈

**Comment by tobiaskohlbau** [comment]
I would like to confirm the latest master version. But it looks like async dns is some kind of broken. DNS-Queries always fail as the timeout is set to zero.

File: plugins/out_azure/azure_conf.c
Comment (nokute78) [reply plugins/out_azure/azure_conf.c]: Thank you.

**Comment by nokute78** [comment]
Currently DNS uses TCP mode as a default.
It may help to use `net.dns.mode UDP` as a UDP mode.

https://github.com/fluent/fluent-bit/blob/4dba97aa7aa032c10ca812da7619d71509a5b070/src/flb_upstream.c#L37-L41

**Comment by tobiaskohlbau** [comment]
> Currently DNS uses TCP mode as a default.
> It may help to use `net.dns.mode UDP` as a UDP mode.
> 
> https://github.com/fluent/fluent-bit/blob/4dba97aa7aa032c10ca812da7619d71509a5b070/src/flb_upstream.c#L37-L41

TCP mode isn't the default anymore. The default is `NULL` which translates to UDP (`dns_global` is also NULL by default):

https://github.com/fluent/fluent-bit/blob/4dba97aa7aa032c10ca812da7619d71509a5b070/src/flb_network.c#L745-L749

I found the issue, net configurations aren't inserted in the plugin if the configmap isn't loaded through:

 ```
/* Load config map */
    ret = flb_output_config_map_set(ins, (void *) ctx);
    if (ret == -1) {
        return NULL;
    }
````

Without the configmap the settings are override with zero values from within the azure plugin. This leads to a connectin timeout of zero which always fails every DNS request.
https://github.com/fluent/fluent-bit/blob/4dba97aa7aa032c10ca812da7619d71509a5b070/plugins/out_azure/azure_conf.c#L185


Should I prepend my changes with another commit to fix this behavior?

**Comment by nokute78** [comment]
Thank you for pointing it.

> Without the configmap the settings are override with zero values from within the azure plugin. This leads to a connectin timeout of zero which always fails every DNS request.

Hmm. I think it is another issue so we need to open an new issue ticket about DNS.

Currently out_azure doesn't support config map...
We may need a patch to support config map first.

**Comment by tobiaskohlbau** [comment]
The interesting part is that the network settings (for e.g. `net.connect_timeout`) are mirrored into a special configmap. This configmap isn't put into context without the plugin itself using configmaps.

See:
https://github.com/fluent/fluent-bit/blob/4dba97aa7aa032c10ca812da7619d71509a5b070/include/fluent-bit/flb_output.h#L679-L686

It's only put in, if another configmap is supported. I presume this breaks every plugin which does not have configmap support.

**Comment by tobiaskohlbau** [comment]
I've built the latest HEAD of this PR and deployed it to my current setup and everything works as expected.

logs

**Review by nokute78 — APPROVED** [review]
Thank you.

**Comment by nokute78** [comment]
@tobiaskohlbau Could you open an issue which you found.
https://github.com/fluent/fluent-bit/pull/3222#issuecomment-894148340

It affects some other output plugins. I'll take care about them.
(e.g.  `connect_timeout` of out_td is also 0)

**Comment by nokute78** [comment]
The network setting issue has been opened as https://github.com/fluent/fluent-bit/issues/3952

**Comment by tobiaskohlbau** [comment]
> The network setting issue has been opened as #3952

Thanks for creating that issue. I'm sorry I forgot to create that issue, was busy at work last week.

**Pull request merged**
